subsys/caf/modules/ble_adv.c

```diff
@@ -721,99 +848,60 @@ static bool handle_ble_peer_operation_event(const struct ble_peer_operation_even
 
 static bool handle_power_down_event(const struct power_down_event *event)
 {
-	int err = 0;
-
 	switch (state) {
-	case STATE_ACTIVE_FAST:
-	case STATE_ACTIVE_SLOW:
-		if (IS_ENABLED(CONFIG_CAF_BLE_ADV_GRACE_PERIOD) &&
-		    (req_grace_period_s > 0)) {
-			err = enter_grace_period();
-		} else {
-			err = ble_adv_stop();
-			if (!err) {
-				state = STATE_OFF;
-				module_set_state(MODULE_STATE_OFF);
-			}
+	case STATE_ACTIVE:
+		if (IS_ENABLED(CONFIG_CAF_BLE_ADV_GRACE_PERIOD) && (req_grace_period_s > 0)) {
+			update_state(STATE_GRACE_PERIOD);
+			break;
 		}
-		break;
+		/* Fall-through. */
 
-	case STATE_DELAYED_ACTIVE_FAST:
-	case STATE_DELAYED_ACTIVE_SLOW:
-	case STATE_ACTIVE_FAST_DIRECT:
-	case STATE_ACTIVE_SLOW_DIRECT:
-		err = ble_adv_stop();
-		if (!err) {
-			state = STATE_OFF;
-			module_set_state(MODULE_STATE_OFF);
-		}
+	case STATE_DELAYED_ACTIVE:
+	case STATE_IDLE:
+		update_state(STATE_OFF);
 		break;
 
-	case STATE_IDLE:
-		state = STATE_OFF;
-		module_set_state(MODULE_STATE_OFF);
+	case STATE_DISABLED:
+		update_state(STATE_DISABLED_OFF);
 		break;
 
 	case STATE_OFF:
 	case STATE_GRACE_PERIOD:
 	case STATE_DISABLED_OFF:
+	case STATE_ERROR:
 		/* No action */
 		break;
 
-	case STATE_DISABLED:
-		state = STATE_DISABLED_OFF;
-		break;
-
 	default:
 		/* Should never happen */
 		__ASSERT_NO_MSG(false);
 		break;
 	}
 
-	if (err) {
-		module_set_state(MODULE_STATE_ERROR);
-	}
-
-	return (state != STATE_OFF) &&
-	       (state != STATE_DISABLED);
+	return !is_module_state_off(state);
```

Comment:
Notice that you will get a different return value now when state will be set to `STATE_DISABLED` or `STATE_DISABLED_OFF`. Is it intended?

Reply:
Yes, actually I think it used to be handled incorrectly. We should return `true` to consume `power_down_event` until we are ready to power down (so when we are not yet in an off state).

Reply:
Keep in mind that if module receives `power_down_event` in `disabled` state, it would go to `disabled off` state before checking this condition.